Live Preview: Introduce the upgrade modal in Premium and WooCommerce themes

okmttdhr opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/79223, https://github.com/Automattic/wp-calypso/pull/83630, https://github.com/Automattic/wp-calypso/pull/84676
https://github.com/Automattic/wp-calypso/pull/84483

## Proposed Changes

This PR introduces the upgrade modal that activates when a user, while live-previewing a theme, clicks on the `Upgrade` button. 

### UI

the modal (for Premium themes)
<img width="1431" alt="Screenshot 2023-12-14 at 15 28 03" src="https://github.com/Automattic/wp-calypso/assets/5287479/3dacd6eb-7167-4de1-bb9f-c715550801c4">

the modal (for WooCommerce themes)
<img width="1435" alt="Screenshot 2023-12-14 at 15 27 23" src="https://github.com/Automattic/wp-calypso/assets/5287479/965f1723-bd86-44ce-a2cb-9fd9bd1d6fb9">

'edit' mode
<img width="1431" alt="Screenshot 2023-12-11 at 16 01 41 copy" src="https://github.com/Automattic/wp-calypso/assets/5287479/93ec16a9-2fe9-4159-8099-5a554049639c">

'view' mode
<img width="450" alt="Screenshot 2023-12-11 at 15 14 05" src="https://github.com/Automattic/wp-calypso/assets/5287479/e676aa7f-7441-472f-a1c1-0a7ad1b3eab5">

### Flow

https://github.com/Automattic/wp-calypso/assets/5287479/bbe73c28-8435-4ddd-9968-e1ac33bebe2c


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Run `install-plugin.sh wpcom-block-editor add/live-preview-upgrade-modal` on your sandbox.
- Sandbox widgets.wp.com and your site.
- Follow the testing instruction in https://github.com/Automattic/wp-calypso/issues/79223

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

okmttdhr committed:
Make ThemeUpgradeModal available from the outside of Calypso

okmttdhr committed:
Add LivePreviewUpgradeModal

okmttdhr committed:
Add handlers to open the modal

okmttdhr commented on client/components/theme-upgrade-modal/index.tsx:
```diff
@@ -55,6 +56,8 @@ interface UpgradeModalContent {
 }
 
 export const ThemeUpgradeModal = ( {
```
I've adjusted the `ThemeUpgradeModal` for its integration in the Site Editor. I'm open to suggestions if there are more optimal methods to use it outside of Calypso.

fushar replied:
Should we add a comment saying that we're not supposed to use Calypso state in this component due to this reason?

okmttdhr committed:
Change the copy to Upgrade

okmttdhr committed:
Hide the border

okmttdhr committed:
Remove the Upgrade action on the notices

okmttdhr commented on apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx:
```diff
@@ -0,0 +1,98 @@
+	}, [ canvasMode ] );
+
+	/**
+	 * Load the SVG sprite for `Gridicon` in `@automattic/components`.
```
I followed the existing code for other apps and could not delve deeper to find a better way. 🤔 

- https://github.com/Automattic/wp-calypso/blob/373449edcf24f6488de2e084b3595b58a8aebfd4/apps/blaze-dashboard/README.md?plain=1#L30-L41
- https://github.com/Automattic/jetpack/blob/5e1e8c83d06ac24ccc88afdcc386c54d61242709/projects/packages/blaze/src/class-dashboard.php#L83-L89
- https://github.com/Automattic/wp-calypso/blob/373449edcf24f6488de2e084b3595b58a8aebfd4/apps/blaze-dashboard/webpack.config.js#L187-L190

fushar replied:
@okmttdhr We can just inline the SVG code. See: https://github.com/Automattic/wp-calypso/pull/82787

cc: @arthur791004 

Sorry for the late review!

okmttdhr replied:
Yes, but I was concerned that we need to discontinue using Gridicons in `client/components/theme-upgrade-modal` (maybe potentially other components in `client/components`). I wanted to keep the component as is as possible. 
But I'm starting to feel it should be acceptable in this specific case.

arthur791004 replied:
>  We can just inline the SVG code

If so, we have to modify the `ThemeUpgradeModal` to use the inline SVG but it seems weird as the component is located under the `client/components` folder and it should use `GridIcon` directly 😅

okmttdhr replied:
I had a try with replacing icons on build time. WDYT? https://github.com/Automattic/wp-calypso/pull/85013/commits/92534977356e9db89ca9129b61818141424903e8

fushar replied:
Sorry maybe I'm missing something but why not just replace gridicons with WP icons in that `ThemeUpgradeModal` directly?

okmttdhr replied:
To be sure, did you see the comment above? 👀 
https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1424813237

fushar replied:
Yes I did. I mean, I don't see why we can't use `@wordpress/icons` in the `ThemeUpgradeModal` if it renders essentially the same thing (a checkmark).

arthur791004 replied:
> I don't see why we can't use @wordpress/icons in the ThemeUpgradeModal if it renders essentially the same thing (a checkmark).

+1. I remember I left the same comment somewhere but I cannot see it 😂

arthur791004 replied:
> How about replacing icons in the modal with the WordPress icon if possible?

Ooops this

okmttdhr replied:
> replace gridicons with WP icons in that ThemeUpgradeModal directly

Thanks, I’ll address it in that way! 👍 

While I agree that it's acceptable in this specific case, I just want to note that my concern was primarily about replacing the original icons (Gridicons), not specifically about whether to implement them inline. https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1424667153

I've now realized that in Calypso, several different icon sets are being used (`@wordpress/icons`, `Gridicons`, `MaterialIcon`). This is not ideal, ![badge](https://img.shields.io/badge/review-imo-orange.svg). Icons are not something interchangeably replaced with different sets even if they appear "similar", and are supposed to be integral to our brand identity.
After things have settled on my end, I’ll create an issue to make it visible.

okmttdhr replied:
📝 p7H4VZ-4JT-p2

autumnfjeld commented:
The opacity or the color in the background here looks odd, like it is too white.    Can a different base color be used? Or something changed so the modal pops more against the background, particularly when the theme beneath is white.  

![image](https://github.com/Automattic/wp-calypso/assets/5665959/45551e17-dd7d-4025-831a-9373672eb255)

okmttdhr committed:
Fix the background color

okmttdhr commented:
> The opacity or the color in the background here looks odd, like it is too white. Can a different base color be used? Or something changed so the modal pops more against the background, particularly when the theme beneath is white.

I fixed it to use the same background color 👍

arthur791004 commented:
hmmm... I can not see the SVG icons. Am I missing something? But the current way to load SVG icons seems to be fragile...

![image](https://github.com/Automattic/wp-calypso/assets/13596067/3bd46653-9ee3-4b1c-ba2d-f4fd7fa77151)

How about replacing icons in the modal with the WordPress icon if possible?

renatho commented on client/my-sites/theme/hooks/use-bundle-settings.tsx:
```diff
@@ -24,11 +24,11 @@ const WooOnPlansIcon = () => (
  */
-const useBundleSettings = ( themeId: string ): BundleSettings | null => {
-	const themeSoftwareSet = useSelector( ( state ) => getThemeSoftwareSet( state, themeId ) );
+export const useBundleSettingsFromThemeSoftwareSet = (
```
Hi @okmttdhr! Thank you for adding me as a reviewer here.
We did a similar thing here in parallel. https://github.com/Automattic/wp-calypso/pull/84886/files#r1422629619

Basically, in my case, I used the `useBundleSettings` getting by the `themeSoftware`, and I created another one with the name `useBundleSettingsByTheme` to get by theme.

I also changed the default export to a named export because I noticed later that it's the way it's used in the majority of the hooks in Calypso.

WDYT? I think it would be good if we apply the same code in both PRs to avoid issues.

okmttdhr replied:
Thank you for letting me know, @renatho! 😄
I will modify my approach to align with your https://github.com/Automattic/wp-calypso/pull/84886, as I prefer `useBundleSettings` if we are retrieving `bundleSettings` from `themeSoftware` on a one-to-one basis.

okmttdhr replied:
📝 I tried to be minimalist in applying #84886: https://github.com/Automattic/wp-calypso/pull/85013/commits/d87ac5485d810bad40c7602a25560ad6a9bdfb07

taipeicoder commented:
Minor detail, but is it an easy change to update the tooltip text as well?

![Screenshot 2023-12-13 at 9 24 42 AM](https://github.com/Automattic/wp-calypso/assets/797888/4b4ee0e8-5ddf-4645-94bd-e6d9901ca6a2)
![Screenshot 2023-12-13 at 9 24 25 AM](https://github.com/Automattic/wp-calypso/assets/797888/9bc28a3d-18c9-4929-90b3-201cd86aabcb)

taipeicoder commented:
@okmttdhr would it be possible to override the keyboard shortcut for saving? <kbd>⌘</kbd><kbd>S</kbd>
Otherwise, it's possible to work around the Upgrade button:

https://github.com/Automattic/wp-calypso/assets/797888/5520c226-9a4a-4104-bb27-1cfdcff9cc13

okmttdhr committed:
Override the keyboard shortcut (⌘S) for saving

okmttdhr committed:
Override the tooltip

okmttdhr committed:
Extract useOverrideSaveButton

okmttdhr committed:
Fix `Failed to execute 'observe' on 'MutationObserver'`

fushar commented:
Wait a minute... So if I made some changes, then click `Upgrade now`, I will be redirected to the checkout, then got redirected back to the Site Editor, but my changes are lost... is this the expected flow? 🥲 

I think so because we currently don't have the mechanism to save without activating (CMIIW).

But this seems weird...

okmttdhr commented:
> Wait a minute... So if I made some changes, then click `Upgrade now`, I will be redirected to the checkout, then got redirected back to the Site Editor, but my changes are lost... is this the expected flow? 🥲
> 
> I think so because we currently don't have the mechanism to save without activating (CMIIW).
> 
> But this seems weird...

Yes, that is expected... （´・ω・｀）.;:…（´・ω...:.;::..（´・;::: .:.;:

fushar commented:
I saw https://github.com/Automattic/wp-calypso/pull/85013/commits/8f097602348e4ec7eaa55a1aaa244db4864de24f but I'm still able to "save" the changes using Cmd-S (it's not being overridden), are you sure it works on your side?

taipeicoder commented:
> > Wait a minute... So if I made some changes, then click `Upgrade now`, I will be redirected to the checkout, then got redirected back to the Site Editor, but my changes are lost... is this the expected flow? 🥲
> > I think so because we currently don't have the mechanism to save without activating (CMIIW).
> > But this seems weird...
> 
> Yes, that is expected, and we have no way to save before activation (yet) ... （´・ω・｀）.;:…（´・ω...:.;::..（´・;::: .:.;:

This seems to be a pretty significant caveat. cc: @autumnfjeld 
Ideally the changes are saved before leaving the Site Editor to the checkout, can we create a spike for this?

fushar commented on apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-previewing-theme.ts:
```diff
@@ -53,9 +52,10 @@ export const usePreviewingTheme = () => {
 		}
 		return;
-	}, [ previewingThemeSlug ] );
+	}, [ previewingThemeId, previewingThemeSlug ] );
```
[nit] I believe `previewingThemeSlug` needs not be a dependency here.

fushar commented:
> hmmm... I can not see the SVG icons. Am I missing something?

Same, I can't see the SVG icons 🥲

fushar commented on apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx:
```diff
@@ -0,0 +1,54 @@
+import './upgrade-modal.scss';
+
+const GRIDICONS_URL =
+	'https://widgets.wp.com/wpcom-block-editor/images/gridicons-47c7fb356fcb2d963681.svg';
```
This loads 404 for me (sandboxed):

<img width="283" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1525580/05502070-9106-4af2-8d8a-90ffb8d0f0c7">

fushar commented on apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts:
```diff
@@ -0,0 +1,116 @@
+					if ( node.nodeType === Node.ELEMENT_NODE ) {
+						const tooltip = ( node as Element ).querySelector( '.components-tooltip' );
+						if ( tooltip ) {
+							tooltip.textContent = __( 'Upgrade now ⌘S', 'wpcom-live-preview' );
```
Can this be something like

```
tooltip.textContent = __( 'Upgrade now', 'wpcom-live-preview' ) + '⌘S';
```

So that it will just reuse the same translation for the Save button?

taipeicoder replied:
I just checked, and LTR locales will have the command on the left side:
![Screenshot 2023-12-13 at 4 18 48 PM](https://github.com/Automattic/wp-calypso/assets/797888/c0983c3e-3637-4b67-857a-911ba4c0fa29)

fushar replied:
(╯°□°)╯︵ ┻━┻

okmttdhr replied:
@taipeicoder @fushar 
I've noticed the command remains on the right side (despite the weird `margin-right`) for the words that have not yet been translated into RTL languages even if I set my language to "العربية". 
<img width="130" alt="Screenshot 2023-12-14 at 12 25 46" src="https://github.com/Automattic/wp-calypso/assets/5287479/4e832f8c-c14b-468a-8796-4d4112032f91">
<img width="334" alt="Screenshot 2023-12-14 at 12 28 29" src="https://github.com/Automattic/wp-calypso/assets/5287479/63bcc469-4cd0-4d73-a10d-b08f59980947">

But the DOM structure seems to be the same.
<img width="590" alt="Screenshot 2023-12-14 at 12 30 02" src="https://github.com/Automattic/wp-calypso/assets/5287479/2d368682-dceb-45b9-9abd-00ee39ec33f1">
<img width="600" alt="Screenshot 2023-12-14 at 12 30 22" src="https://github.com/Automattic/wp-calypso/assets/5287479/324663ab-001e-4e53-96e6-e89983487ee1">

Can you help me find out how it's implemented (possibly) on the Gutenberg side? 
We may be able to implement it like `textContent = isRtl ? command + text : text + command` as we override the copy anyway, but I want to understand the current implementation to know if we can just use the current DOM structure without doing `textContent = isRtl ? command + text : text + command`. 

Or would it be okay if we take care of RTL in later iterations? What do you think?

fushar replied:
I found that the shortcut text and its placement is computed dynamically. See:

https://github.com/WordPress/gutenberg/blob/2fb3c57bcadc675f9c6745a99d957b3924e39b71/packages/edit-site/src/components/save-button/index.js#L96

My suggestion for now is to NOT include / leave out the shortcut from the tooltip, due to the complexity. What do you think?

okmttdhr replied:
> leave out the shortcut from the tooltip

@fushar In that case, the tooltip displays the same text as the button label, making it redundant imo. 😃 
What do you think about hiding the tooltip entirely, then?

taipeicoder replied:
I'm in favor of leaving out the shortcut, but keeping the tooltip for accessibility purposes. Even though it might be redundant 🙂

okmttdhr replied:
I addressed this in https://github.com/Automattic/wp-calypso/pull/85013/commits/3af814b75852537d84da08d3f53853af25b4c759 🙂

okmttdhr committed:
Replace Gridicons with WordPress icons

okmttdhr commented:
> I saw [8f09760](https://github.com/Automattic/wp-calypso/commit/8f097602348e4ec7eaa55a1aaa244db4864de24f) but I'm still able to "save" the changes using Cmd-S (it's not being overridden), are you sure it works on your side?

There seems to be regression. I'll fix it, and thanks for being patient 😣

okmttdhr committed:
Fix regression of the keyboard shortcut (⌘S) for saving

okmttdhr committed:
Remove unnecessary deps

okmttdhr committed:
Add a comment about usage of ThemeUpgradeModal

okmttdhr committed:
Remove the shortcut text from the tooltip

See https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1426166628

okmttdhr committed:
Replace Gridicons directly in ThemeUpgradeModal

okmttdhr commented:
I have addressed all the critical feedback. I'd appreciate it if you take a look at it again 🙏 🙏 @Automattic/lego 

- Changed to use WP icons to show them outside of Calypso. https://github.com/Automattic/wp-calypso/pull/85013/commits/be7ac439d3e44c1198f56cdd3af1620a9167f118
- Overrode the default behavior of the keyboard shortcut and the tooltip. https://github.com/Automattic/wp-calypso/pull/85013/commits/b169c3c094f1693522f202f89b27592ed659a067 https://github.com/Automattic/wp-calypso/pull/85013/commits/5ea0caf651601be9562aa5e80c339a190b6d8df1

okmttdhr committed:
Apply #84886 to useBundleSetting

https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1424374607

okmttdhr committed:
Replace Gridicons directly in ThemeUpgradeModal (Remove the loader)

taipeicoder commented on apps/wpcom-block-editor/webpack.config.js:
```diff
@@ -58,6 +59,9 @@ function getWebpackConfig(
 			...webpackConfig.plugins.filter(
 				( plugin ) => plugin.constructor.name !== 'DependencyExtractionWebpackPlugin'
 			),
+			new webpack.DefinePlugin( {
```
@okmttdhr could you explain why do we need this change? Thanks!

okmttdhr replied:
This is needed for `import`-ing `ThemeUpgradeModal`, which is directly due to the use of NODE_DEBUG in the package called [util](https://www.npmjs.com/package/util). I'll add a comment about it 👍

taipeicoder commented on apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts:
```diff
@@ -0,0 +1,127 @@
+				setIsThemeUpgradeModalOpen( true );
+			}
+		};
+		document.addEventListener( 'keydown', ( e: KeyboardEvent ) => {
```
Should this be just:
```suggestion
		document.addEventListener( 'keydown',  overrideSaveButtonKeyboardShortcut );
```

okmttdhr replied:
🤦

taipeicoder commented on client/components/theme-upgrade-modal/index.tsx:
```diff
@@ -69,11 +77,15 @@ export const ThemeUpgradeModal = ( {
-	const bundleSettings = useBundleSettings( slug );
+	// Currently, it always get the first software set. In the future, the whole applications can be enhanced to support multiple ones.
+	const firstThemeSoftwareSet = themeSoftwareSet?.[ 0 ];
+	const bundleSettings = useBundleSettings( firstThemeSoftwareSet?.slug );
```
Should this be `useBundleSettingsByTheme`?

okmttdhr replied:
No, as I decided to refrain from using Calypso state this time. 🙂 
Related: https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1421971735, https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1424374607, https://github.com/Automattic/wp-calypso/pull/84886#discussion_r1426350658

okmttdhr replied:
CC: @renatho

taipeicoder reviewed (approved):
Tested and it works as intended 👍 
Left some minor questions.

okmttdhr committed:
Use the defined function

okmttdhr committed:
Add a comment about build settings change

fushar commented:
Looks like there are conflicts... Could you resolve them and I'll test this again first thing tomorrow 😄

okmttdhr commented:
I'll wait for @fushar's review before merging 🙂

fushar reviewed (approved):
Working well.

Thanks for the hard work!

okmttdhr commented:
>https://github.com/Automattic/wp-calypso/pull/85013#issuecomment-1853398541
> > > Wait a minute... So if I made some changes, then click `Upgrade now`, I will be redirected to the checkout, then got redirected back to the Site Editor, but my changes are lost... is this the expected flow? 🥲
> > > I think so because we currently don't have the mechanism to save without activating (CMIIW).
> > > But this seems weird...
> > 
> > 
> > Yes, that is expected, and we have no way to save before activation (yet) ... （´・ω・｀）.;:…（´・ω...:.;::..（´・;::: .:.;:
> 
> This seems to be a pretty significant caveat. cc: @autumnfjeld Ideally the changes are saved before leaving the Site Editor to the checkout, can we create a spike for this?

In reference to the previous discussions about saving functionality in https://github.com/Automattic/wp-calypso/issues/81119, I believe the following same principles apply here:

> The current behavior is already clear, if we think the BTP as a truly "Preview" experience (quoting @autumnfjeld's words). We also have communicated this behavior clearly in two places:
https://github.com/Automattic/wp-calypso/issues/81119#issuecomment-1831307892

Given that we can still perceive it as "a true Preview experience", and we have the notices indicating that users cannot save changes before upgrading (`You can try out your own style customizations, which will only be saved if you upgrade and activate this theme.`), 
We might want to retain user editings during the same flow (= 'Live Preview in Premium and WooCommerce themes') at least. But, if we decide to do a spike, we may want to aim to think about these complexities https://github.com/Automattic/wp-calypso/issues/81119#issuecomment-1831307892. This may include considering the saving feature in Live Preview comprehensively to ensure a consistent and non-contradictory user experience.

taipeicoder commented:
> We might want to retain user editings during the same flow (= 'Live Preview in Premium and WooCommerce themes') at least.

Yes. I think this user scenario is more nuanced than https://github.com/Automattic/wp-calypso/issues/81119, in terms of (1) we are not providing a Save button to the user, and (2) I don't think it's a good UX for users to change the theme → upgrade → redo the changes → activate the theme.

The wording "You can try out your own style customizations, which will only be saved if you upgrade and activate this theme." implies the order of trying out changes → upgrade → activate, which is not technically correct since the changes will be gone after upgrading.

That said, we don't necessarily need to save when users go to checkout. Just need to improve the UX so that the second point mentioned above can be further avoided 🙂

okmttdhr commented:
It would be worthwhile to see how many users make what changes before upgrading their plan.
